File: src/features/typing/data/go/coin_change.go
package main

import "math"

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		dp[i] = math.MaxInt
	}
	for i := 1; i <= amount; i++ {
		for _, coin := range coins {
			if coin <= i && dp[i-coin] != math.MaxInt && dp[i-coin]+1 < dp[i] {
				dp[i] = dp[i-coin] + 1
			}
		}
	}
	if dp[amount] == math.MaxInt {
		return -1
	}
	return dp[amount]
}

File: src/features/typing/data/dart/coin_change.dart
class Solution {
  int coinChange(List<int> coins, int amount) {
    int max = 1 << 30;
    List<int> dp = List.filled(amount + 1, max);
    dp[0] = 0;
    for (var i = 1; i <= amount; i++) {
      for (var coin in coins) {
        if (coin <= i && dp[i - coin] != max && dp[i - coin] + 1 < dp[i]) {
          dp[i] = dp[i - coin] + 1;
        }
      }
    }
    if (dp[amount] == max) {
      return -1;
    }
    return dp[amount];
  }
}
